// ng-image-slider/esm5/lib/ng-image-slider.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-image-slider.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgImageSliderComponent } from './ng-image-slider.component';
import { SliderCustomImageComponent } from './slider-custom-image/slider-custom-image.component';
import { SliderLightboxComponent } from './slider-lightbox/slider-lightbox.component';
import { NgImageSliderService } from './ng-image-slider.service';
import * as ɵngcc0 from '@angular/core';
var NgImageSliderModule = /** @class */ (function () {
    function NgImageSliderModule() {
    }
NgImageSliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgImageSliderModule });
NgImageSliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgImageSliderModule_Factory(t) { return new (t || NgImageSliderModule)(); }, providers: [
        NgImageSliderService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgImageSliderModule, { declarations: function () { return [NgImageSliderComponent,
        SliderCustomImageComponent,
        SliderLightboxComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgImageSliderComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgImageSliderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    NgImageSliderComponent,
                    SliderCustomImageComponent,
                    SliderLightboxComponent
                ],
                providers: [
                    NgImageSliderService
                ],
                exports: [NgImageSliderComponent]
            }]
    }], function () { return []; }, null); })();
    return NgImageSliderModule;
}());
export { NgImageSliderModule };
