// @cloudinary/angular-5.x/esm5/lib/cloudinary-transformation.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var CloudinaryTransformationDirective = /** @class */ (function () {
    function CloudinaryTransformationDirective(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    CloudinaryTransformationDirective.prototype.getAttributes = /**
     * @return {?}
     */
    function () {
        return this.el.nativeElement.attributes;
    };
    /** @nocollapse */
    CloudinaryTransformationDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
CloudinaryTransformationDirective.ɵfac = function CloudinaryTransformationDirective_Factory(t) { return new (t || CloudinaryTransformationDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
CloudinaryTransformationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CloudinaryTransformationDirective, selectors: [["cl-transformation"]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CloudinaryTransformationDirective, [{
        type: Directive,
        args: [{
                selector: 'cl-transformation'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return CloudinaryTransformationDirective;
}());
export { CloudinaryTransformationDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CloudinaryTransformationDirective.prototype.el;
}
