// @cloudinary/angular-5.x/esm2015/lib/cloudinary-transformation.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class CloudinaryTransformationDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    getAttributes() {
        return this.el.nativeElement.attributes;
    }
}
CloudinaryTransformationDirective.ɵfac = function CloudinaryTransformationDirective_Factory(t) { return new (t || CloudinaryTransformationDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
CloudinaryTransformationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CloudinaryTransformationDirective, selectors: [["cl-transformation"]] });
/** @nocollapse */
CloudinaryTransformationDirective.ctorParameters = () => [
    { type: ElementRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CloudinaryTransformationDirective, [{
        type: Directive,
        args: [{
                selector: 'cl-transformation'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    CloudinaryTransformationDirective.prototype.el;
}
